### PR TITLE
Reduction dimension moved to correct place

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**

### CODE REVIEW

File: lib/TPP/RewriteToBatchReduceGEMM.cpp [hunk 1]
```diff
@@ -106,20 +107,23 @@ static LogicalResult checkVNNIAccessPatterns(linalg::LinalgOp linalgOp) {
     maps.push_back(map.getMinorSubMap(2));
   }
 
+  llvm::errs() << "compressed dims maps\n";
   SmallVector<AffineMap> compressedDimMaps = compressUnusedDims(maps);
   using MapList = ArrayRef<ArrayRef<AffineExpr>>;
+  for (auto dimMap : compressedDimMaps)
```
Comment (chelini) [inline lib/TPP/RewriteToBatchReduceGEMM.cpp:113]: same here.

File: lib/TPP/ToBlockLayoutAndBack.cpp [hunk 2]
```diff
@@ -507,27 +510,50 @@ mlir::linalgx::packVNNIMatmulOp(RewriterBase &rewriter,
...
-      ArrayRef<mlir::utils::IteratorType>{mlir::utils::IteratorType::parallel,
-                                          mlir::utils::IteratorType::parallel,
-                                          mlir::utils::IteratorType::reduction,
-                                          mlir::utils::IteratorType::parallel,
-                                          mlir::utils::IteratorType::parallel,
-                                          mlir::utils::IteratorType::reduction,
-                                          mlir::utils::IteratorType::reduction},
-      /*doc=*/"", /*libraryCall=*/"");
+  linalg::GenericOp replacementOp;
+  if (matmulOp.getInputs()[1].getType().cast<ShapedType>().getRank() == 4) {
```
Comment (chelini) [inline lib/TPP/ToBlockLayoutAndBack.cpp:518]: why do we have to differentiate here?

**Review by chelini — COMMENTED** [review]
please also add a test.

File: lib/TPP/RewriteToBatchReduceGEMM.cpp [hunk 3]
```diff
@@ -106,20 +107,23 @@ static LogicalResult checkVNNIAccessPatterns(linalg::LinalgOp linalgOp) {
     maps.push_back(map.getMinorSubMap(2));
   }
 
+  llvm::errs() << "compressed dims maps\n";
```
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/RewriteToBatchReduceGEMM.cpp:110]: you can use the LLVM_DBG macro thing, if this is something that will help development in the future, not just a quick local print

File: lib/TPP/ToBlockLayoutAndBack.cpp (hunk 2 above)
Comment (KavithaTipturMadhu) [reply lib/TPP/ToBlockLayoutAndBack.cpp:518]: The maps and iterators to linalg.generic differ for operands of different shapes (i.e., when shape is 4d, generic is 7 dimensional and maps are as per the code, as compared to 3d shape obtained when using tiling factor of 1,1).

Comment (chelini) [reply lib/TPP/ToBlockLayoutAndBack.cpp:520]: Now I am worried. If we start adding a case for each combination, this does not scale.

Comment (chelini) [reply lib/TPP/ToBlockLayoutAndBack.cpp:520]: If tiling is a 1,1 it tiles along parallel dimensions and indeed this seems the case looking at the affine map, I don't think we need to differentiate, we should look only at the last [r, r, p, p, r] dimensions.

Comment (KavithaTipturMadhu) [reply lib/TPP/ToBlockLayoutAndBack.cpp:525]: Added a todo for now.

**Comment by KavithaTipturMadhu** [comment]
Updated PR with a test case for untiled version of matmul @chelini. Review?

**Review by chelini — APPROVED** [review]
Thanks, I am fine merging as is to allow making progress. Please add an issue and feel free to assign it to me.

**Pull request merged**